### PR TITLE
feat(gui): show system drives by default with warnings

### PR DESCRIPTION
**Pull request by Shou**
Instead of hiding system drives, they are now shown by default but with
added warnings upon attempts to select or flash a system drive.

- Red 'Continue' button in drive-list widget when system drive selected
- Alert icons added to system drives
- DriveInfoModal added
  - Used to warn users before flashing a system drive
  - Used to display drive details
- Unsafe mode removed

See: https://github.com/resin-io/etcher/issues/888
Changelog-Entry: Show system drives by default with warnings.

### CODE REVIEW

**Comment by Shou** [comment]
Here are some screenshots

Before flashing, if your drive's a system drive


The drive details modal


Drive selector modal
image

**Comment by WasabiFan** [comment]
Some comments/questions:
- When does the "drive details" modal appear? Is that a link below the drive selection step?
- The "All data on your system will be lost" text sounds awkward and isn't actually correct. I think that the word "destroy(ed)" should be included somewhere and that the extreme "all" should be removed or qualified.
- The text alignment in the last image feels wrong; it seems like that text -- the line above the button -- should be centered.
- The button text "change" feels odd... a more direct word might be better.

**Comment by Shou** [comment]
@WasabiFan 
* So the drive details modal appears when you click the drive mount point (below the circled 2nd step); the text is accompanied by an alert triangle when it's a system drive, clicking it will show exactly what's 'wrong' through the information presented by the modal.
* Yeah we may have to tweak the wording a bit here, "all" is technically wrong (in most cases). Maybe "this will result in an unbootable system and data loss"?
* Indeed, I actually centred it locally a jiffy ago.
* I think the idea behind the wording was "change [drive]". The "change" button is going to open (but doesn't yet) the drive selector modal.

**Comment by lurch** [comment]
Ping @konmouz for him to compare this to his designs :)

"this will result in an unbootable system and data loss" isn't technically true, because although it may be a "system drive", it may not necessarily be the "boot drive" - the extra subtlety is dependent on https://github.com/resin-io-modules/drivelist/issues/126

I agree the text "All data on your system will be lost" is misleading, but I think the "All data on your system drive will be lost" proposed in #888 _might_ be okay.

"The "change" button is going to open (but doesn't yet)" - does that mean that this PR isn't ready for merging yet?

Is the drive details modal also available for removable drives? And if so could you include a screenshot of that too please?

File: lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    <div class="modal-selected-drive-description">
+      {{ ::modal.selection.getDrive().description }}
+
+    </div>
+    <div class="modal-selected-drive-name">
+      {{ ::modal.selection.getDrive().name }} - {{ ::modal.selection.getDrive().size | gigabyte | number:1 }} GB
+
+
+      <span class="label label-warning"
+          ng-show="modal.selection.isDriveLargeEnough(modal.selection.getDrive()) && !modal.selection.isDriveLocked(modal.selection.getDrive()) && !modal.selection.isDriveSizeRecommended(modal.selection.getDrive())">
```
Comment (lurch) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: Now that the logic on this line...

File: lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          <i class="glyphicon glyphicon-warning-sign"></i>
+          NOT RECOMMENDED</span>
+
+      <span class="label label-danger"
+        ng-hide="modal.selection.isDriveLargeEnough(modal.selection.getDrive())">
+          <i class="glyphicon glyphicon-resize-small"></i>
+          TOO SMALL FOR IMAGE</span>
+
+      <span class="label label-danger"
+        ng-show="modal.selection.isDriveLocked(modal.selection.getDrive()) && modal.selection.isDriveLargeEnough(modal.selection.getDrive())">
```
Comment (lurch) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: ...and the logic on this line is shared between `drive-selector-modal.tpl.html` and `drive-info-modal.tpl.html`, perhaps this logic of categorising drives into "OKAY" (no warning label), "NOT RECOMMENDED", "TOO SMALL FOR IMAGE", "LOCKED" and "SYSTEM DRIVE" ought to be refactored out into the model itself (or some other shared JS file), rather than having the logic embedded inside HTML templates?

Comment (lurch) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: Also @jviotti it looks like the current logic is dictating that if a drive is too small, and it's locked, then the warning shown is "TOO SMALL FOR IMAGE".
IMHO the fact that it's locked, out to be displayed in preference to the fact that it's too small? Moving the categorisation logic outside of the template file will make it easier to make this logic more consistent, without this current mish-mash of some of the warning labels here using `ng-show` and some of them using `ng-hide`.

File: lib/gui/models/store.js [hunk 3]
```diff
@@ -90,9 +89,17 @@ const storeReducer = (state, action) => {
...
+        return !drive.system;
+      });
+
+      const selectedDevice = newState.getIn([ 'selection', 'drive' ]);
 
-        const drive = _.first(action.data);
+      // There are non-system drives; set one as the selected drive.
+      if (!selectedDevice && noSystemDrives.length >= 1) {
+
+        const drive = _.first(noSystemDrives);
```
Comment (lurch) [inline lib/gui/models/store.js]: Um, is this auto-selecting the first removable drive, if there are multiple drives available?
IMHO Etcher shouldn't do that, it should just stick with the current behaviour of only auto-selecting a drive if it's the **only** removable drive available.

File: lib/gui/models/store.js [hunk 4]
```diff
@@ -90,9 +89,17 @@ const storeReducer = (state, action) => {
 
       const newState = state.set('availableDrives', Immutable.fromJS(action.data));
 
-      if (action.data.length === 1) {
+      // Non-system drives
+      const noSystemDrives = _.filter(action.data, (drive) => {
```
Comment (lurch) [inline lib/gui/models/store.js]: In preparation for the addition of "Boot drive" detection (amongst other filtering), could this `noSystemDrives` be renamed to `removableDrives` ?

File: lib/gui/pages/main/controllers/main.js [hunk 5]
```diff
@@ -65,4 +66,22 @@ module.exports = function(
     return this.shouldDriveStepBeDisabled() || !SelectionStateModel.hasDrive();
   };
 
+  /**
+   * @summary Open a drive info modal
+   * @function
+   * @public
+   *
+   * @returns {Boolean} whether the flash step should be disabled
```
Comment (lurch) [inline lib/gui/pages/main/controllers/main.js]: Copy'n'Paste error?

File: lib/gui/components/drive-selector/controllers/drive-selector.js [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 const _ = require('lodash');
 
-module.exports = function($q, $uibModalInstance, DrivesModel, SelectionStateModel, WarningModalService) {
+module.exports = function($q, $uibModalInstance, DrivesModel, SelectionStateModel, SettingsModel, WarningModalService) {
```
Comment (lurch) [inline lib/gui/components/drive-selector/controllers/drive-selector.js]: Is this change needed?

**Comment by lurch** [comment]
I hope you don't think I'm being overly harsh, I just have a very good eye for detail ;-)

It's nice to see the removal of the "Unsafe Mode" :+1:

**Comment by lurch** [comment]
P.S. Looks like the CI tests are failing because the unit tests need to be updated.

File: lib/gui/components/drive-selector/controllers/drive-selector.js (hunk 6 above)
Comment (Shou) [inline lib/gui/components/drive-selector/controllers/drive-selector.js]: Odd, I think `npm test` should've warned me about that?

File: lib/gui/pages/main/controllers/main.js (hunk 5 above)
Comment (Shou) [inline lib/gui/pages/main/controllers/main.js]: Indeed! I'll get that fixed.

File: lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      <span class="label label-danger"
+        ng-show="modal.selection.getDrive().system">
+          <i class="glyphicon glyphicon-hdd"></i>
+          SYSTEM DRIVE</span>
+
+    </div>
+  </div>
+</div>
+
+<div class="modal-footer"
```
Comment (jviotti) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: Why does this `DriveInfoModal` component does two things (displaying information and asking for confirmation)? There should be two different component to do different tasks, however, I think you should re-use the`warning-modal` here instead of creating new one.

File: lib/gui/components/drive-info-modal/services/drive-info-modal.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+   *
+   * @param {Object} options - options
+   * @param {String} options.description - danger message
+   * @param {String} options.confirmationLabel - confirmation button text
+   * @param {String} options.rejectionLabel - rejection button text
+   * @fulfil {Boolean} - whether the user accepted or rejected the warning
+   * @returns {Promise}
+   *
+   * @example
+   * WarningModalService.display({
```
Comment (jviotti) [inline lib/gui/components/drive-info-modal/services/drive-info-modal.js]: The example calls `WarningModalService.display` instead of `DriveInfoModalService.display`. In any case, we should re-use `WarningModalService` (extending it if necessary) instead of copy-pasting it here.

File: lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    <p>{{ ::modal.options.description }}</p>
+  </div>
+  <div class="modal-selected-drive">
+    <div class="modal-selected-drive-description">
+      {{ ::modal.selection.getDrive().description }}
+
+    </div>
+    <div class="modal-selected-drive-name">
+      {{ ::modal.selection.getDrive().name }} - {{ ::modal.selection.getDrive().size | gigabyte | number:1 }} GB
+
```
Comment (jviotti) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: There are two blank lines here.

File: lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+      <span class="label label-warning"
```
Comment (jviotti) [inline lib/gui/components/drive-info-modal/templates/drive-info-modal.tpl.html]: This whole set of labels are being copy-pasted from the drive selector. If we're showing them in two places now, we should extract this logic so it can be re-used between both, otherwise we risk having them out of sync.

File: lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html [hunk 11]
```diff
@@ -45,7 +50,15 @@ <h4 class="list-group-item-heading">{{ drive.description }} - {{ drive.size | gi
 </div>
 
 <div class="modal-footer">
+  <span ng-show="modal.state.getDrive().system" class="modal-footer-warning">
```
Comment (jviotti) [inline lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html]: I'd rename `modal-footer-warning` to something like `modal-footer-label`, since there is nothing in its definition that is specific to a "warning" element:

```css
.modal-drive-selector-modal .modal-footer-warning {
  padding: 8px;
  display: inline-block;
}
```

File: lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html [hunk 12]
```diff
@@ -45,7 +50,15 @@ <h4 class="list-group-item-heading">{{ drive.description }} - {{ drive.size | gi
...
+      Etcher has identified this as a system drive
+  </span>
   <button class="button button-primary button-block"
+    ng-class="{ 'label-danger': modal.state.getDrive().system, 'button-primary': !modal.state.getDrive().system }"
```
Comment (jviotti) [inline lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html]: `button-primary` is both in `ngClass` and in the element's `class` attribute. I believe your intention was to but the "danger" class in `ngClass`, such that is gets added when the drive is a system drive, right?

(BTW, the button doesn't go red here for system drives)

screenshot 2016-12-11 20 32 50

Comment (jviotti) [inline lib/gui/components/drive-selector/templates/drive-selector-modal.tpl.html]: We should be re-using the `isSystemDrive()` function you added to `SelectionStateModel` instead of `modal.state.getDrive().system`.

File: lib/gui/models/selection-state.js [hunk 13]
```diff
@@ -169,6 +169,28 @@ SelectionStateModel.service('SelectionStateModel', function(DrivesModel) {
...
+   *   device: '/dev/disk2',
+   *   name: 'My Drive',
+   *   size: 123456789,
+   *   protected: true,
+   *   system: true
+   * })) {
+   *   console.log('This drive is a system drive!');
+   * }
+   */
+  this.isSystemDrive = (drive) => {
```
Comment (jviotti) [inline lib/gui/models/selection-state.js]: This function should contain unit tests.

File: lib/gui/models/selection-state.js [hunk 14]
```diff
@@ -169,6 +169,28 @@ SelectionStateModel.service('SelectionStateModel', function(DrivesModel) {
...
+    return drive.system;
```
Comment (jviotti) [inline lib/gui/models/selection-state.js]: Can we cast to boolean, just to be on the safe side?

```js
return Boolean(drive.system);
```

File: lib/gui/models/store.js [hunk 15]
```diff
@@ -90,9 +89,17 @@ const storeReducer = (state, action) => {
...
+        return !drive.system;
+      });
+
+      const selectedDevice = newState.getIn([ 'selection', 'drive' ]);
 
-        const drive = _.first(action.data);
+      // There are non-system drives; set one as the selected drive.
+      if (!selectedDevice && noSystemDrives.length >= 1) {
```
Comment (jviotti) [inline lib/gui/models/store.js]: This new functionality should be unit tested as well.

File: lib/gui/modules/drive-scanner.js [hunk 16]
```diff
@@ -53,13 +53,7 @@ driveScanner.factory('DriveScannerService', (SettingsModel) => {
...
       });
 
-      if (SettingsModel.get('unsafeMode')) {
-        return drives;
-      }
-
-      return _.reject(drives, (drive) => {
-        return drive.system;
-      });
+      return drives;
```
Comment (jviotti) [inline lib/gui/modules/drive-scanner.js]: This new functionality should be unit tested.

File: lib/gui/pages/main/controllers/flash.js [hunk 17]
```diff
@@ -26,10 +26,50 @@ module.exports = function(
...
+   * FlashController.flashImageToDrive('rpi.img', {
+   *   device: '/dev/disk2',
+   *   description: 'Foo',
+   *   size: 99999,
+   *   mountpoint: '/mnt/foo',
+   *   system: false
+   * });
+   */
+  this.flashImageToDriveModal = (image, drive) => {
+    if (drive.system) {
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: Lets reuse `isSystemDrive` here.

File: lib/gui/pages/main/controllers/flash.js [hunk 18]
```diff
@@ -26,10 +26,50 @@ module.exports = function(
...
+        icon: 'alert',
+        title: 'ATTENTION',
+        header: 'FLASH SYSTEM DRIVE?',
+        description: 'All data on your system will be lost.'
+      }).then((userAccepted) => {
+        if (userAccepted) {
+          this.flashImageToDrive(image, drive);
+        } else {
+
+          // DriveSelectionController.openDriveSelector();
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: This comment, and empty `else` clause should be removed.

File: lib/gui/pages/main/controllers/flash.js [hunk 19]
```diff
@@ -26,10 +26,50 @@ module.exports = function(
...
+   * @example
+   * FlashController.flashImageToDrive('rpi.img', {
+   *   device: '/dev/disk2',
+   *   description: 'Foo',
+   *   size: 99999,
+   *   mountpoint: '/mnt/foo',
+   *   system: false
+   * });
+   */
+  this.flashImageToDriveModal = (image, drive) => {
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: This variable name implies that `this.flashImageToDriveModal` represents the modal itself, and not a function that opens something else (the `DriveInfoModal` component). Maybe we can rename to something like `askFlashConfirmation` or something like that?

Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: This function uses `this.flashImageToDrive()`, however its declared before it. Can we move its declaration below?

File: lib/gui/pages/main/controllers/main.js [hunk 20]
```diff
@@ -65,4 +66,22 @@ module.exports = function(
...
+   *
+   * @example
+   * if (MainController.shouldFlashStepBeDisabled()) {
+   *   console.log('The flash step should be disabled');
+   * }
+   */
+  this.openDriveInfo = () => {
+    return DriveInfoModalService.display({
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/main.js]: How does this function know which drive to open?

Comment (jviotti) [inline lib/gui/pages/main/controllers/main.js]: Oh, I see it reads from `SelectionStateModel` internally. A better design would be to decouple `DriveInfoModal` from `SelectionStateModel` and pass the drive to it from here.

File: lib/gui/pages/main/templates/main.tpl.html [hunk 21]
```diff
@@ -67,7 +67,11 @@
           <div class="step-selection-text"
             ng-class="{
               'text-disabled': main.shouldDriveStepBeDisabled()
-            }">{{ main.selection.getDrive().name }} - {{ main.selection.getDrive().size | gigabyte | number:1 }} GB</div>
+            }"
+            ng-click="main.openDriveInfo()">{{ main.selection.getDrive().name }} - {{ main.selection.getDrive().size | gigabyte | number:1 }} GB
+            <i class="glyphicon glyphicon-alert"
+              ng-show="main.selection.getDrive().system"></i>
```
Comment (jviotti) [inline lib/gui/pages/main/templates/main.tpl.html]: Lets reuse `isSystemDrive` here.

File: lib/gui/pages/main/controllers/main.js [hunk 22]
```diff
@@ -19,6 +19,7 @@
 module.exports = function(
   SelectionStateModel,
   DrivesModel,
+  DriveInfoModalService,
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/main.js]: The main page should declare the `drive-info-modal` component as a dependency before attempting to use it.

File: lib/gui/modules/drive-scanner.js [hunk 23]
```diff
@@ -32,7 +32,7 @@ const driveScanner = angular.module(MODULE_NAME, [
   require('../models/settings')
```
Comment (jviotti) [inline lib/gui/modules/drive-scanner.js]: If this component doesn't depend on `SettingsModel`  anymore, then it should remove it from its dependencies.

File: lib/gui/pages/main/controllers/flash.js [hunk 24]
```diff
@@ -26,10 +26,50 @@ module.exports = function(
   FlashErrorModalService,
   ErrorService,
   OSNotificationService,
-  OSWindowProgressService
+  OSWindowProgressService,
+  DriveInfoModalService
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: Make sure the main page declares `DriveInfoModal` as a dependency before using it.

File: lib/gui/pages/main/controllers/main.js [hunk 25]
```diff
@@ -65,4 +66,22 @@ module.exports = function(
...
+   * @public
+   *
+   * @returns {Boolean} whether the flash step should be disabled
+   *
+   * @example
+   * if (MainController.shouldFlashStepBeDisabled()) {
+   *   console.log('The flash step should be disabled');
+   * }
+   */
+  this.openDriveInfo = () => {
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/main.js]: Maybe `showDriveInformation` is a better name?

File: lib/gui/pages/main/controllers/main.js [hunk 26]
```diff
@@ -65,4 +66,22 @@ module.exports = function(
...
+    return DriveInfoModalService.display({
+      title: 'Drive details'
+    });
```
Comment (jviotti) [inline lib/gui/pages/main/controllers/main.js]: Lets use the `ErrorService` here to catch any potential error.

**Comment by jviotti** [comment]
This is very nice, however its a huge PR, and it will be difficult to review appropriately. I propose to split this PR into the following changes, so that we can review it more easily, and get it merged one step at a time:

- Show a "SYSTEM DRIVE" badge if the drive is a system drive. This would include the `isSystemDrive` implementation in `SelectionStateModel`.
- Turn the "Continue" button to red in the drive selector and show the warning label if the drive is a system drive.

(Notice that at this point, we are not even showing the system drives by default, so this changes only apply if the user clicked "unsafe mode") 

- Remove the "unsafe mode" option and add the confirmation modal before "flash".
- Show the "drive information" dialog.

***

You can do them in order and keep rebasing this big one once every step is merged.

**Comment by jviotti** [comment]
I'm closing this PR since drive selection has been re-designed on v2.
